plenum/server/node.py

```diff
@@ -1444,16 +1517,81 @@ def postRecvTxnFromCatchup(self, ledgerId: int, txn: Any):
             rh = self.reqHandler
         return rh
 
+    # TODO: should be renamed to `post_all_ledgers_caughtup`
     def allLedgersCaughtUp(self):
-        if self.ledgerManager.lastCaughtUpPpSeqNo > 0:
-            # TODO: currently we use the same ppSeqNo for all instances
-            for replica in self.replicas:
-                replica.caught_up_till_pp_seq_no(self.ledgerManager.lastCaughtUpPpSeqNo)
+        last_caught_up_3PC = self.ledgerManager.last_caught_up_3PC
+        if compare_3PC_keys(self.master_last_ordered_3PC,
+                            last_caught_up_3PC) > 0:
+            self.master_replica.caught_up_till_pp_seq_no(last_caught_up_3PC)
+            logger.debug('{} caught up till {}'.format(self,
+                                                       last_caught_up_3PC))
+
+        # TODO: Ensure no more catchups are required, the catchup needs to
+        # communicate number of txns caught up, use `catchupTill`
+        if self.is_catchup_needed():
+            logger.debug('{} needs to catchup again'.format(self))
+            self.start_catchup()
+        else:
+            logger.debug('{} does not need any more catchups'.format(self))
+            self.no_more_catchups_needed()
+
+    def is_catchup_needed(self) -> bool:
+        """
+        Check if received a quorum of view change done messages and if yes check if caught up till the
+        Check if all requests ordered till last prepared certificate
+        Check if last catchup resulted in no txns
+        """
+        if self.caught_up_for_current_view():
+            logger.info('{} is caught up for the current view {}'.format(self, self.viewNo))
+            return False
+        logger.debug('{} is not caught up for the current view {}'.format(self, self.viewNo))
+        if self.has_ordered_till_last_prepared_certificate() and \
```

Comment:
Is it sufficient condition to finish catch-up?
Hoe can we finish catch-up if `caught_up_for_current_view()` failed, that is we have different state?